Sources/Vapor/Response/Response.swift

```diff
@@ -71,33 +117,41 @@ public final class Response: CustomStringConvertible {
         }
 
         func encode<E>(_ encodable: E, using encoder: ContentEncoder) throws where E : Encodable {
-            var body = self.response.body.byteBufferAllocator.buffer(capacity: 0)
-            try encoder.encode(encodable, to: &body, headers: &self.response.headers)
-            self.response.body = .init(buffer: body, byteBufferAllocator: self.response.body.byteBufferAllocator)
+            try self.response.responseBox.withLockedValue { box in
+                var body = box.body.byteBufferAllocator.buffer(capacity: 0)
+                try encoder.encode(encodable, to: &body, headers: &box.headers)
+                box.body = .init(buffer: body, byteBufferAllocator: box.body.byteBufferAllocator)
+            }
         }
 
         func decode<D>(_ decodable: D.Type, using decoder: ContentDecoder) throws -> D where D : Decodable {
-            guard let body = self.response.body.buffer else {
-                throw Abort(.unprocessableEntity)
+            try self.response.responseBox.withLockedValue { box in
+                guard let body = box.body.buffer else {
+                    throw Abort(.unprocessableEntity)
+                }
+                return try decoder.decode(D.self, from: body, headers: box.headers)
             }
-            return try decoder.decode(D.self, from: body, headers: self.response.headers)
         }
 
         func encode<C>(_ content: C, using encoder: ContentEncoder) throws where C : Content {
             var content = content
             try content.beforeEncode()
-            var body = self.response.body.byteBufferAllocator.buffer(capacity: 0)
-            try encoder.encode(content, to: &body, headers: &self.response.headers)
-            self.response.body = .init(buffer: body, byteBufferAllocator: self.response.body.byteBufferAllocator)
+            try self.response.responseBox.withLockedValue { box in
+                var body = box.body.byteBufferAllocator.buffer(capacity: 0)
+                try encoder.encode(content, to: &body, headers: &box.headers)
+                box.body = .init(buffer: body, byteBufferAllocator: box.body.byteBufferAllocator)
+            }
         }
 
         func decode<C>(_ content: C.Type, using decoder: ContentDecoder) throws -> C where C : Content {
-            guard let body = self.response.body.buffer else {
-                throw Abort(.unprocessableEntity)
+            try self.response.responseBox.withLockedValue { box in
+                guard let body = box.body.buffer else {
+                    throw Abort(.unprocessableEntity)
+                }
+                var decoded = try decoder.decode(C.self, from: body, headers: box.headers)
+                try decoded.afterDecode()
+                return decoded
             }
```

Comment:
The `try decoded.afterDecode()` should be outside?
```suggestion
            var decoded = try self.response.responseBox.withLockedValue { box in
                guard let body = box.body.buffer else {
                    throw Abort(.unprocessableEntity)
                }
                return try decoder.decode(C.self, from: body, headers: box.headers)
            }
            try decoded.afterDecode()
            return decoded
```

Reply:
Functionally it doesn't make a difference, the code has the same result, it just releases the lock at a different time. For a sequential block like this it shouldn't matter

Reply:
Right the lock release was what I was slightly worrying about. It won't matter much but I think it's still better to release it when you can.